pkg/cmd/experimental/syncgroups/schema1/ldapinterface.go

```diff
@@ -0,0 +1,183 @@
+
+	"k8s.io/kubernetes/pkg/util"
+
+	"github.com/go-ldap/ldap"
+
+	"github.com/openshift/origin/pkg/auth/ldaputil"
+	"github.com/openshift/origin/pkg/cmd/experimental/syncgroups/interfaces"
+	"github.com/openshift/origin/pkg/cmd/server/api"
+)
+
+// NewLDAPInterface builds a new LDAPInterface using a schema-appropriate config
+func NewLDAPInterface(clientConfig ldaputil.LDAPClientConfig,
+	userAttributeDefiner ldaputil.LDAPUserAttributeDefiner,
+	groupQueryTemplate api.LDAPQueryTemplate,
+	userQueryTemplate api.LDAPQueryFromAttributeTemplate) LDAPInterface {
+	return LDAPInterface{
+		clientConfig:         clientConfig,
+		userAttributeDefiner: userAttributeDefiner,
+		groupQueryTemplate:   groupQueryTemplate,
+		userQueryTemplate:    userQueryTemplate,
+		groupNameAttribute:   []string{},
+		cachedUsers:          make(map[string]*ldap.Entry),
+		cachedGroups:         make(map[string]*ldap.Entry),
+	}
+}
+
+var _ interfaces.LDAPMemberExtractor = &LDAPInterface{}
+var _ interfaces.LDAPGroupGetter = &LDAPInterface{}
+var _ interfaces.LDAPGroupLister = &LDAPInterface{}
+
+// LDAPInterface extracts the member list of an LDAP group entry from an LDAP server
+// with first-class LDAP entries for groups. The LDAPInterface is *NOT* thread-safe.
+// The LDAPInterface satisfies:
+// - LDAPMemberExtractor
+// - LDAPGroupGetter
+// - LDAPGroupLister
+type LDAPInterface struct {
+	// clientConfig holds LDAP connection information
+	clientConfig ldaputil.LDAPClientConfig
+	// groupQueryTemplate holds the information necessary to make an LDAP query for a specific
+	// first-class group entry on the LDAP server
+	groupQueryTemplate api.LDAPQueryTemplate
+	// userQueryTemplate holds the information necessary to make an LDAP query for a specific
+	// first-class user entry on the LDAP server, given a first-class group entry with member records
+	// stored in an attribute.
+	userQueryTemplate api.LDAPQueryFromAttributeTemplate
+	// groupNameAttribute is the list of LDAP group entry attributes that correspond to the OpenShift
+	// Group Name field.
+	groupNameAttribute []string
+	// LDAPUserAttributeDefiner defines the values corresponding to OpenShift Identities in LDAP entries
+	// by using a deterministic mapping of LDAP entry attributes to OpenShift Identity fields
+	userAttributeDefiner ldaputil.LDAPUserAttributeDefiner
+
+	// cachedGroups holds the result of group queries for later reference, indexed on group UID
+	// e.g. this will map an LDAP group UID to the LDAP entry returned from the query made using it
+	cachedGroups map[string]*ldap.Entry
+	// cachedUsers holds the result of user queries for later reference, indexed on user UID
+	// e.g. this will map an LDAP user UID to the LDAP entry returned from the query made using it
+	cachedUsers map[string]*ldap.Entry
+}
+
+// ExtractMembers returns the LDAP member entries for a group specified with a ldapGroupUID
+func (e *LDAPInterface) ExtractMembers(ldapGroupUID string) (members []*ldap.Entry, err error) {
```

Comment:
Don't name the return args in the implementations.  Its good to do in the interface definitions, but not the impls